library/java/org/cronvoy/UserAgent.java

```diff
@@ -0,0 +1,97 @@
+package org.cronvoy;
+
+import android.content.Context;
+import android.content.pm.PackageInfo;
+import android.content.pm.PackageManager;
+import android.content.pm.PackageManager.NameNotFoundException;
+import android.os.Build;
+import java.util.Locale;
+
+/** Constructs a User-Agent string. */
+final class UserAgent {
+
+  private static final Object sLock = new Object();
+
+  private static final int VERSION_CODE_UNINITIALIZED = 0;
+  private static int sVersionCode = VERSION_CODE_UNINITIALIZED;
+
+  /**
+   * Constructs a User-Agent string including application name and version, system build version,
+   * model and Id, and Cronet version.
+   *
+   * @param context the context to fetch the application name and version from.
+   * @return User-Agent string.
+   */
+  static String from(Context context) {
```

Comment:
As far as I can tell, there is no direct test for this. However, many tests uses this `@Rule`:

https://chromium.googlesource.com/chromium/src/+/master/components/cronet/android/test/javatests/src/org/chromium/net/CronetTestRule.java#205

The hope here is this will get implicitly tested by the virtue of using that test rule. Not awesome.

Reply:
Yeah, maybe we can create an issue and link it in this PR. What do you think?

Reply:
Issue created: https://github.com/envoyproxy/envoy-mobile/issues/1437